Use TORCH_CHECK instead of std::runtime_error in stack.h and ivalue.h

janeyx99 opened the pull request:
TORCH_CHECK will preserve the stacktrace for when TORCH_CPP_SHOW_STACKTRACES=1, whereas std::runtime_error will not.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145280

janeyx99 committed:
Use TORCH_CHECK instead of std::runtime_error in stack.h and ivalue.h

[ghstack-poisoned]

malfet commented on aten/src/ATen/core/stack.h:
```diff
@@ -103,9 +103,7 @@ inline void drop(Stack* stack, size_t n) {
-  if (stack.empty()) {
-    throw std::runtime_error("pop() called on empty stack");
-  }
+  TORCH_CHECK(not stack.empty(), "pop() called on empty stack");
```
```suggestion
  TORCH_CHECK(!stack.empty(), "pop() called on empty stack");
```

Skylion007 replied:
@malfet This is unfortunately valid in C++17...

malfet replied:
@Skylion007 , well, according to VC++2019 it's `not true` (pun intended) :)

janeyx99 committed:
Update on "Use TORCH_CHECK instead of std::runtime_error in stack.h and ivalue.h"

TORCH_CHECK will preserve the stacktrace for when TORCH_CPP_SHOW_STACKTRACES=1, whereas std::runtime_error will not.




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
